subprojects/diagnostics/src/test/groovy/org/gradle/api/tasks/diagnostics/PropertyReportTaskTest.groovy

```diff
@@ -57,4 +57,36 @@ class PropertyReportTaskTest extends Specification {
         1 * renderer.addProperty("prop", "value")
         1 * renderer.addProperty("properties", "{...}")
     }
+
+    def "when a specific property is requested, only that property is requested"() {
+        when:
+        task.setProperty("prop")
+        task.generate(project)
+
+        then:
+        1 * project.properties >> ["prop": "value", "properties": "prop"]
+        1 * renderer.addProperty("prop", "value")
+        0 * renderer.addProperty("properties", "{...}")
+    }
+
+    def "does not show contents of the properties property when requested"() {
+        when:
+        task.setProperty("properties")
+        task.generate(project)
+
+        then:
+        1 * project.properties >> ["prop": "value", "properties": "prop"]
+        1 * renderer.addProperty("properties", "{...}")
+    }
+
+
+    def "when a specific property is requested, but isn't found"() {
+        when:
+        task.setProperty("not-found")
+        task.generate(project)
+
+        then:
+        1 * project.properties >> ["properties": "prop"]
+        1 * renderer.addProperty("not-found", null)
+    }
```

Comment:
This is an ancient test class that was migrated from JMock. I think all implementation should go. I've rewritten it with your extra tests. PTAL and replace the entire class:

```groovy
class PropertyReportTaskTest extends Specification {

    @Rule
    final TestNameTestDirectoryProvider temporaryFolder = new TestNameTestDirectoryProvider(getClass())

    private ProjectInternal project
    private PropertyReportRenderer renderer
    private PropertyReportTask task

    def setup() {
        project = Mock()
        renderer = Mock()
        task = TestUtil.create(temporaryFolder).task(PropertyReportTask.class)
        task.setRenderer(renderer)
    }

    def "passes each project property to renderer"() {
        given:
        project.properties >> ["b": "value2", "a": "value1"]

        when:
        task.generate(project)

        then:
        1 * renderer.addProperty("a", "value1")
        1 * renderer.addProperty("b", "value2")
    }

    def "uses placeholder for rendering 'properties' property"() {
        given:
        project.properties >> ["prop": "value", "properties": "prop"]

        when:
        task.generate(project)

        then:
        1 * renderer.addProperty("prop", "value")
        1 * renderer.addProperty("properties", "{...}")
    }

    def "can show a single property"() {
        given:
        project.properties >> ["b": "value2", "a": "value1"]
        task.property = "b"

        when:
        task.generate(project)

        then:
        1 * renderer.addProperty("b", "value2")
        0 * renderer.addProperty("a", "value2")
    }

    def "uses placeholder for rendering 'properties' property even if it's selected via PropertyReportTask.property"() {
        given:
        project.properties >> ["prop": "value", "properties": "prop"]
        task.property = 'properties'

        when:
        task.generate(project)

        then:
        0 * renderer.addProperty("prop", "value")
        1 * renderer.addProperty("properties", "{...}")
    }

    def "passes unavailable properties to renderer"() {
        given:
        project.properties >> [:]
        task.property = 'not-found'

        when:
        task.generate(project)

        then:
        1 * renderer.addProperty("not-found", null)
    }
}

```